test/cpp/tensorexpr/test_loopnest.cpp

```diff
@@ -1528,5 +1528,56 @@ void testOuterLoopVectorization() {
   ASSERT_EQ(dynamic_cast<For*>(for_body->front()), nullptr);
 }
 
+namespace {
+
+std::string constantUpperBoundLoopIR(int upper_bound_val) {
+  KernelScope kernel_scope;
+  ExprHandle upper_bound(upper_bound_val);
+  Tensor* A = Compute(
+      "A", {{upper_bound, "x"}}, [&](const VarHandle& x) { return x * 2; });
+  LoopNest l({A});
+  Stmt* s = l.root_stmt();
+  std::vector<For*> loops = l.getLoopStmtsFor(A);
+  Stmt* unrolled = nullptr;
+  l.unroll(loops[0], &unrolled);
+  std::ostringstream oss;
+  oss << *unrolled;
+  return oss.str();
+}
+
+} // namespace
+
+void testUnroll() {
+  const std::string actual = constantUpperBoundLoopIR(3);
+  const std::string& verification_pattern =
+      R"IR(
+# CHECK: A[0] = 0;
+# CHECK: A[1] = 2;
+# CHECK: A[2] = 4)IR";
+
+  torch::jit::testing::FileCheck().run(verification_pattern, actual);
+}
+
+void testUnrollEmpty() {
+  const std::string actual = constantUpperBoundLoopIR(0);
+  const std::string& verification_pattern = R"IR(
+# CHECK-NOT: A[
+  )IR";
+
+  torch::jit::testing::FileCheck().run(verification_pattern, actual);
+}
+
+void testNoUnroll() {
+  KernelScope kernel_scope;
+  VarHandle upper_bound("N", kInt);
+  Tensor* A = Compute(
+      "A", {{upper_bound, "x"}}, [&](const VarHandle& x) { return x * 2; });
+  LoopNest l({A});
+  Stmt* s = l.root_stmt();
+  std::vector<For*> loops = l.getLoopStmtsFor(A);
+  Stmt* unrolled = nullptr;
+  ASSERT_THROWS_WITH(l.unroll(loops[0], &unrolled), "non-constant loop");
+}
```

Comment:
Missing a few test cases we should cover:
* Unrolling a loop with multiple stmts in the body (because we don't have loop fusion yet, you'll need to build the IR manually. E.g.
```
Stmt* s = For::make(VarHandle("i", kInt), 0, 4, Block::Make({
  Store::make(...);
  Store::make(...);
});
```
* Unrolling a loop whose body() Block contains a VarMapping (again, will need to manually create the IR since these are inserted by transforms).
* Unrolling an inner nested loop.
* Unrolling an outer nested loop (feel free to do these last two together if thats easier).

Reply:
Added most of the tests you suggested - not sure what you're looking for the Block which contains a VarMapping. Could you explain a bit more?

Reply:
Sure, the Block stmt has a set of variable mappings which are written out at the top of the block's statements (think variables in Basic or Pascal). We should probably test something like this:
```
  VarHandle e("e", kInt);
  auto f = For::make(
      x,
      0,
      kTotalSize,
      Block::make({e.node(), new IntImm(7)},
                  {Store::make(a_buf, {x}, e),
                   Store::make(b_buf, {x}, e + 1)}));
```

Which expands to something like this:
```
for (int x = 0; x < kTotalSize; ++i) {
  int e = 7;
  a_buf[x] = e;
  b_buf[x] = e + 1;
}
```

This is a bit different because you don't want to replace the definition of e multiple times.

Reply:
Done.